core/partitioning/shape_analysis.cpp

```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
   return ivalue_map;
 }
 
+torch::jit::Node* getUpstreamCastNode(torch::jit::Value* val) {
+  std::queue<torch::jit::Value*> q;
+  q.push(val);
+  std::unordered_set<torch::jit::Node*> visited;
+  while (!q.empty()) {
+    auto cur_val = q.front();
+    q.pop();
+    auto node = cur_val->node();
+    if (node->kind().toQualString() == std::string("aten::to")) {
```

Comment:
May need an additional check to ensure that the `aten::to` schema is valid for `dtype` insertion, as some of these schemas do not take an integer `dtype` at all, for example:

- `aten::to(Tensor(a) self, bool non_blocking=False, bool copy=False) -> Tensor(b|a)` 
- `aten::to(Tensor(a) self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)`
- `aten::to(Tensor(a) self, Tensor other, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)`

A check could be something like an additional `&&` with

```cpp
(node->inputs()[1]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType) ||
(node->inputs()[2]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType)
```

Reply:
Hey @gs-olive Any reproducer for this? 
What I'm not sure about is that for getUpstreamNode() function when we pass in a int32 value will the first `cast` node be the cast node that casts this value to int64? If that's the case, then we don't need this check. 
In other words, is it possible that the first `cast` node involving the passed value is to cast some other value? If the first `cast` node is not the cast node that casts to int64, will the second `cast` node be what we want?

Reply:
Hi @bowang007 - yes, this did throw an error on my end, when compiling the [BART](https://huggingface.co/docs/transformers/model_doc/bart) model with the changes made in [this PR](https://github.com/pytorch/TensorRT/pull/1416), to fix an unrelated issue. The linked PR uses the second `aten::to` listed above, which does not take an integer dtype, so inserting the integer dtype there throws an error. Specifically, these lines in the PR cause the issue: https://github.com/pytorch/TensorRT/blob/9e7822aa842ee50316f1cac5d9d841ed04353c15/core/lowering/passes/device_casting.cpp#L28-L29

My thought was that we do need this check you have, but maybe it should be more stringent - something like:

```cpp
    if ((node->kind().toQualString() == std::string("aten::to")) &&
        ((node->inputs()[1]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType) ||
        (node->inputs()[2]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType))) {
```

Reply:
@bowang007 Please let me know what you think about the above

Reply:
@gs-olive I got your point now, let me update this part.